pytential/utils.py

```diff
@@ -32,4 +34,61 @@ def sort_arrays_together(*arys, key=None):
     """
     return zip(*sorted([x for x in zip(*arys)], key=key))
 
+
+def chop(expr, tol):
+    """Given a symbolic expression, remove all occurences of numbers
+    with absolute value less than a given tolerance and replace floating
+    point numbers that are close to an integer up to a given relative
+    tolerance by the integer.
+    """
+    nums = expr.atoms(sym.Number)
+    replace_dict = {}
+    for num in nums:
+        if float(abs(num)) < tol:
+            replace_dict[num] = 0
+        else:
+            new_num = float(num)
+            if abs((int(new_num) - new_num)/new_num) < tol:
+                new_num = int(new_num)
+            replace_dict[num] = new_num
+    return expr.xreplace(replace_dict)
+
+
+def lu_solve_with_expand(L, U, perm, b):
+    """Given an LU factorization and a vector, solve a linear
+    system with intermediate results expanded to avoid
+    an explosion of the expression trees
+
+    :param L: lower triangular matrix
+    :param U: upper triangular matrix
+    :param perm: permutation matrix
+    :param b: column vector to solve for
+    """
+    def forward_substitution(L, b):
+        n = len(b)
+        res = sym.Matrix(b)
+        for i in range(n):
+            for j in range(i):
+                res[i] -= L[i, j]*res[j]
+            res[i] = (res[i] / L[i, i]).expand()
+        return res
+
+    def backward_substitution(U, b):
+        n = len(b)
+        res = sym.Matrix(b)
+        for i in range(n-1, -1, -1):
+            for j in range(n - 1, i, -1):
+                res[i] -= U[i, j]*res[j]
+            res[i] = (res[i] / U[i, i]).expand()
+        return res
+
+    def permuteFwd(b, perm):
+        res = sym.Matrix(b)
+        for p, q in perm:
+            res[p], res[q] = res[q], res[p]
+        return res
+
+    return backward_substitution(U,
+            forward_substitution(L, permuteFwd(b, perm)))
```

Comment:
Some things here look very non-flake8. How come the linter didn't yell?

Reply:
Looks like `N802` is disabled in setup.cfg because of functions like `S, D, Sp, Dp`